pkg/bgp/controller/controller.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package controller abstracts the MetalLB service IP allocator controller.
+package controller
+
+import (
+	"os"
+
+	bgpconfig "github.com/cilium/cilium/pkg/bgp/config"
+	bgplog "github.com/cilium/cilium/pkg/bgp/log"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/option"
+
+	metallballoc "go.universe.tf/metallb/pkg/allocator"
+	metallbctl "go.universe.tf/metallb/pkg/controller"
+)
+
+// New creates a new BGP MetalLB controller.
+func New() *Controller {
+	logger := &bgplog.Logger{Entry: log}
+	client := &client{k8s.Client()}
+
+	c := &metallbctl.Controller{
+		Client: client,
+		IPs:    metallballoc.New(),
+	}
+
+	f, err := os.Open(option.Config.BGPConfigPath)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to open BGP config file")
+	}
+	defer f.Close()
+
+	config, err := bgpconfig.Parse(f)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to parse BGP configuration")
+	}
+	c.SetConfig(logger, config)
+
+	return &Controller{
+		Controller: c,
+		logger:     logger,
+	}
+}
+
+// Controller represents the BGP controller. It integrates Cilium with the
+// MetalLB logic for allocate LB IP addresses for service objects of type
+// LoadBalancer.
+type Controller struct {
+	*metallbctl.Controller
```

Comment:
I've renamed it to `Manager` because I agree that we shouldn't co-opt the `Controller` notion of Cilium and ultimately cause confusion. 

Good point on the subscriber vs "manager". They actually don't make sense as separate pieces and really should be just one object. I think it turned out this way as an artifact of how I started, but I never revisited the structure as the code evolved. Thanks for the close inspection. :rocket: